script/hooks/migration.js

```diff
@@ -1,5 +1,5 @@
 export const migrateWorld = async () => {
-    const schemaVersion = 4.0;
+    const schemaVersion = 4.1;
```

Comment:
I would be just so much better and simpler to equate model version and the system's version!
But I only noticed after pushing the PR: `schemaVersion` is just a double, not a "semver"-like string or object.

What I did sort of works, but I now see how terribly hackish and brittle it is.
I believe it would be better to use a string and compare versions splitting on dots (`.`). The downside is that we would have to add special handling for migrations when `schemaVersion` was still a double.

What do you think?

Reply:
Yes, I agree. I had to read up on Foundry best practices here, and from what I understand Foundry doesn't care. The Dev Wiki suggests a best practice of semVer for system versioning and we have that already. 

I think, seeing that we have only used integers so far when updating schemaVersions (thus no sorting conflict e.g. 3.10 < 3.9) the better approach here would be to go through the migrations and consolidate them, then add a conditional to migrate anything below 4 (current), and update to new schemaVersion system.

I also suggest we remove settings field for schemaVersion in the system's settings. Alternately, we can set `config=true` dependant on an environment variable in `/src` (when I finally get around to reorganizing the project files). User should at any rate not be provided easy access to that field as it does not serve them any good to mess with the schemaVersion (and those who want to intentionally can easily do so using console).

Finally, I think we should check system version in absence of a schemaVersion (easily done). And the schemaVersion (which should continue to exist as its own independent variable) does not need to be a string (adds complexity) instead either an `Int` or `BigInt` that follows system version e.g. `majmip`, `mamip` or `mmp`. I have no preference here really, but I think `majmip` is an easy to understand format (next release would then look like `400300`.

Reply:
My previous statement does not stand anymore. 

I have been giving this some thought lately, and the conclusion I have arrived at is that @jimorie has (unsurprisingly) handled this correctly. Whenever we touch migration we should do a major version increment. This is because backwards compatibility is always breaking when we touch migration. `template.json` can be manipulated without touching migration, as has been done earlier. So I think our requirement should be Major version increment whenever `migration.js` is edited to migrate data.

This then removes any necessity of using a double, and we can stick to integers. Seeing as we have used a double previously we should build in logic to convert worldSchemaVersion to an integer rounded down (since worldSchemaVersion is an exposed setting, it's not inconceivable that some users have messed with it in some way). I also think it might be prudent to make sure that worldSchemaVersion can never exceed the size of system version.

Reply:
How about something like this:
```js
	const schemaVersion = 5;
	const systemVersion = Number(
		game.system.data.version.replace(/([0-9])[\s\S\w]+/i, "$1") //Get the first whole integer in system version.
	);
	const trueWorldSchemaVersion = Number(
		game.settings.get("forbidden-lands", "worldSchemaVersion") || 0
	); //Moved to a separate variable due to this being an exposed setting that can be set to integer, float or string. This also sets 0 if it evaluates to NaN. 
	const worldSchemaVersion =
		trueWorldSchemaVersion <= systemVersion
			? trueWorldSchemaVersion
			: systemVersion; //Intention here is to prevent worldSchema from being larger than systemversion preventing a migration and potentially "corrupting" the World.
```